zapcore/sampler.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+const (
+	_numLevels        = _maxLevel - _minLevel
+	_countersPerLevel = 4096
+)
+
+func newCounters() *counters {
+	return &counters{}
+}
+
+type counters [_numLevels][_countersPerLevel]atomic.Uint64
+
+func (c *counters) Inc(lvl Level, key string) uint64 {
+	return c.get(lvl, key).Inc()
+}
+
+func (c *counters) Reset(lvl Level, key string) {
+	c.get(lvl, key).Store(0)
+}
+
+func (c *counters) get(lvl Level, key string) *atomic.Uint64 {
+	i := lvl - _minLevel
+	j := fnv32a(key) % _countersPerLevel
+	return &c[i][j]
+}
+
+// fnv32a, adapted from "hash/fnv", but without a []byte(string) alloc
+func fnv32a(s string) uint32 {
+	const (
+		offset32 = 2166136261
+		prime32  = 16777619
+	)
+	hash := uint32(offset32)
+	for i := 0; i < len(s); i++ {
+		hash ^= uint32(s[i])
+		hash *= prime32
+	}
+	return hash
+}
+
+// Sample creates a facility that samples incoming entries.
+func Sample(fac Facility, tick time.Duration, first, thereafter int) Facility {
+	return &sampler{
+		Facility:   fac,
+		tick:       tick,
+		counts:     newCounters(),
+		first:      uint64(first),
+		thereafter: uint64(thereafter),
+	}
+}
+
+type sampler struct {
+	Facility
+
+	tick       time.Duration
+	counts     *counters
+	first      uint64
+	thereafter uint64
+}
+
+func (s *sampler) With(fields []Field) Facility {
+	return &sampler{
+		Facility:   s.Facility.With(fields),
+		tick:       s.tick,
+		counts:     s.counts,
+		first:      s.first,
+		thereafter: s.thereafter,
+	}
+}
+
+func (s *sampler) Check(ent Entry, ce *CheckedEntry) *CheckedEntry {
+	if !s.Enabled(ent.Level) {
+		return ce
+	}
+
+	if n := s.counts.Inc(ent.Level, ent.Message); n > s.first {
+		if n == s.first+1 {
```

Comment:
Depends on which correct behaviour we're talking about I think. We have a few options:
1. current option of `s.first+1` which I don't think has any benefits
2. `s.first` which will start the timer as soon as we're about to start dropping and will handle my above example better.
3. `n == 1`

Let's consider another example: we're logging every 0.5 seconds, and the parameters are `s.first = 2`, `s.tick = time.Second` `s.thereafter = 100` (or some high number).

Option 3 would log every message (which seems like the ideal behaviour), option 2 would drop every 3rd message, option 1 would drop every 3rd and 4th message.

Option 3 will also start a timer every `s.tick` for every log message, which seems really bad. I think Option 2 is a good compromise between performance and sampling.